File: merge-sort/merge-sort.py
#! /usr/bin/env python

from random import random
from math import floor
from time import time
from array import array

def get_index(i):
    return floor(random() * i)


def generate_array(size):
    arr = array("I", [])
    for i in range(1, size + 1):
        index = get_index(i)
        arr.insert(index, i)
    return arr


def get_left_and_right(arr):
    extra = 0 if len(arr) % 2 == 0 else 1
    mid = floor(len(arr) / 2)
    left = arr[0:mid + extra]
    right = arr[mid + extra:]
    return left, right


def order_array(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while (j >= 0 and arr[j] > key):
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr


def merge_left_and_right(Left, Right):
    res_len = len(Left) + len(Right)
    res = array("I", [])
    left_index = 0
    right_index = 0
    for i in range(0, res_len):
        if right_index >= len(Right):
            res = res + Left[left_index:]
            break
        if left_index >= len(Left):
            res = res + Right[right_index:]
            break
        if Left[left_index] < Right[right_index]:
            res.append(Left[left_index])
            left_index += 1
        else:
            res.append(Right[right_index])
            right_index += 1
    return res


def compare_arrays(subject, expected):
    for i in range(0, len(subject)):
        if subject[i] != expected[i]:
            return False
    return True


def main():
    size = 30000

    # Generate array
    gen_start = time()
    arr = generate_array(size)
    gen_end = time()

    sort_start = time()

    # Get Left and Right
    left, right = get_left_and_right(arr)

    # Sort Left and Right (insertion sort)
    left = order_array(left)
    right = order_array(right)

    # Merge Left and Right
    merged = merge_left_and_right(left, right)

    sort_end = time()

    # Eval sorted array with an expected result
    expected = array("I", [])
    for i in range(1, size + 1):
        expected.append(i)
    are_equal = compare_arrays(merged, expected)

    # Output the results
    print(f"1. Time to generate array: {floor((gen_end - gen_start) * 1000)} ms")
    if not are_equal:
        print("2. Result array is NOT sorted as expected")
    else:
        print("2. Result array is sorted as expected")
        print(f"3. Time to merge sort the array: {floor((sort_end - sort_start) * 1000)} ms")

main()

File: merge-sort/merge-sort.js
const generateArray = (size) =>
{
    const arr = []
    for (let i = 1; i <= size; i++) {
        const index = Math.floor(Math.random() * i)
        arr.splice(index, 0, i)
    }
    return arr
}

const getLeftAndRight = (arr) =>
{
    const extra = arr % 2 == 0 ? 0 : 1 // extra when arr.length is odd
    const mid = Math.floor(arr.length / 2) - 1
    const left = arr.slice(0, mid + extra)
    const right = arr.slice(mid + extra)
    return { left, right }
}

const sortArray = (arr) =>
{
    for (let i = 1; i < arr.length; i++) {
        const key = arr[i];
        let j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
    return arr
}

const mergeLeftAndRight = (left, right) =>
{
    const resultLength = left.length + right.length
    let result = []
    let leftIndex = 0
    let rightIndex = 0
    for (let k = 0; k < resultLength; k++) {
        if (rightIndex >= right.length) {
            result = result.concat(left.slice(leftIndex))
            break
        }
        if (leftIndex >= left.length) {
            result = result.concat(right.slice(rightIndex))
            break
        }
        if (left[leftIndex] < right[rightIndex]) {
            result[k] = left[leftIndex]
            leftIndex++
        } else {
            result[k] = right[rightIndex]
            rightIndex++
        }
    }
    return result
}

const compareArrays = (subject, expected) =>
{
    for (let i = 0; i < subject.length; i++) {
        if (subject[i] != expected[i]) return false
    }
    return true
}

const main = () =>
{
    const size = 30_000

    // Generate array
    const genStart = new Date().getTime()
    const arr = generateArray(size)
    const genEnd = new Date().getTime()

    const sortStart = new Date().getTime()

    // Get Left and Right halves of the array
    const { left, right } = getLeftAndRight(arr)

    // Sort Left and Right (with insertionSort)
    sortArray(left)
    sortArray(right)

    // Merge Left and Right
    const merged = mergeLeftAndRight(left, right)

    const sortEnd = new Date().getTime()

    // Eval sorted array with an expected result
    const expected = []
    for (let i = 1; i <= size; i++) expected.push(i)
    const areEqual = compareArrays(merged, expected)

    // Output the results
    console.log("1. Time to generate array: ", genEnd - genStart, "ms")
    if (! areEqual) {
        console.log("2. Result array is NOT sorted as expected")
    } else {
        console.log("2. Result array is sorted as expected")
        console.log("3. Time to merge sort the array: ", sortEnd - sortStart, "ms")
    }
}

main()
